include/boost/gil/image_processing/morphology.hpp

```diff
@@ -0,0 +1,131 @@
+//
+// Copyright 2019 Ayush Bansal <abansal1008@gmail.com>
+//
+// Use, modification and distribution are subject to the Boost Software License,
+// Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+//
+#ifndef BOOST_GIL_IMAGE_PROCESSING_MORPHOLOGY_HPP
+#define BOOST_GIL_IMAGE_PROCESSING_MORPHOLOGY_HPP
+
+#include <boost/assert.hpp>
+#include <boost/gil.hpp>
+
+namespace boost { namespace gil {
+
+// SE refers to Structuring Element
+// Result type after convolution with the structuring element
+enum class morphology_operator{
+    erode,
+    dilate
+};
+
+namespace detail {
+
+template <typename SrcView, typename DstView, typename StructElement>
+void convolve_with_struct_element(SrcView const &src_view, DstView &dst_view,
+    StructElement const &struct_elem, std::ptrdiff_t src_y, std::ptrdiff_t src_x,
+    morphology_operator const &op)
+{
+    // Template argument validation
+    gil_function_requires<ImageViewConcept<SrcView>>();
+    gil_function_requires<MutableImageViewConcept<DstView>>();
+
+    using source_channel_t = typename channel_type<SrcView>::type;
+
+    // Initial min and max values to be replaced after apply morphplogical operator
+    auto min = (std::numeric_limits<source_channel_t>::max());
+    auto max = (std::numeric_limits<source_channel_t>::min());
+
+    std::size_t struct_elem_height = struct_elem.size();
+    std::size_t struct_elem_width = struct_elem[0].size();
+
+    for(std::ptrdiff_t struct_elem_y = 0; struct_elem_y < struct_elem_height; ++struct_elem_y)
+    {
+        // The cooresponding value to be read from src_view is calculated as
+        // (src_y + struct_elem_y - struct_elem_height/2, src_x + struct_elem_x - struct_elem_width/2)
+        // NOTE: This calculation is only valid when the struct_elem dimensions are odd.
+        auto y_to_read = src_y + struct_elem_y - struct_elem_height/2;
+        if(y_to_read < 0 || y_to_read >= src_view.height())
+            continue;
+        typename SrcView::x_iterator src_it = src_view.row_begin(y_to_read);
+
+        for(std::ptrdiff_t struct_elem_x = 0; struct_elem_x < struct_elem_width; ++struct_elem_x)
+        {
+            // We need to consider only those positions which have 1 in Structuring Element
+            if(struct_elem[struct_elem_x][struct_elem_y] == 1)
```

Comment:
Loose comment, a while ago I personally started appreciating the early exits and continuations as recommended by https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code, this also helps to avoid too deep indentations what is a good idea for long names of templated types, etc.

Reply:
I too prefer early exits and continuations. But, for this case I this way seemed more appropriate to me. The only reason being this flow is more coherent with the natural flow of the algorithm and for a new reader it would be more intuitive. And aldo, there is not much deep indentation in this part of code. However, if you still feel if I should use early continuation here, I can do that?

Reply:
It's fine as it is. The number of nested levels is not large to make it a problem. I pointed at the issues to be aware as a tool to avoid code drift towards the right too much :)